Sprint 16

Peppa-Akk opened the pull request:
Sprint 16
- Add logic of category;
- Implemented CoreData to category;
- Category logic is written on the MVVM;
- Add category to Trackers;
- Add Onboard View;
- Add Keychain property to check for the first entrance in application.

Peppa-Akk committed:
Update UI:
- Add scroll; - Add CollectionView; - Configure CollectionView; - Add new view (Schedule); - Configure Schedule View.  Start with CoreData: - Add to project data model (.cdatamodeld file); - Configure data model (entity and relationship); - Init Persistent Container. 

Peppa-Akk committed:
Add subtitle for schedule button

Peppa-Akk committed:
update CoreDataStores:
- add new tracker to Core Data;
- apply category relationship for tracker;
- add new record to Core Data;
- delete record (with id and date);

Fix date logic (now use start day);
Fix present Event View.

Peppa-Akk committed:
sprint 16:
- Add logic of category;
- Implemented CoreData to category;
- Category logic is written on the MVVM;
- Add category to Trackers;
- Add Onboard View;
- Add Keychain property to check for the first entrance in application.

DanilaG commented on Tracker/SplashViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    private let storage = AuthStorage()
+    
+    lazy var pages: [UIViewController] = {
+        let firstPage = PageForOnboardingViewController(page: PageVC.first)
+        let secondPage = PageForOnboardingViewController(page: PageVC.second)
```
Было бы супер, передавать изменяемые параметры страницы в ините её view controller. В данном случае ими могут выступать картинки и текста. Тогда, если понадобиться добавить новую страницу - то это можно будет сделать значительно проще, так как все данные находятся в одном месте + не требуется менять файл самого `PageForOnboardingViewController`)

DanilaG commented on Tracker/SplashViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,120 @@
+        guard let viewControllerIndex = pages.firstIndex(of: viewController) else {
+            return nil
+        }
+        let nextIndex = viewControllerIndex + 1
+        
+        guard nextIndex < pages.count else {
+            return pages.first
+        }
```
Можно воспользоваться модульной арифметикой: `let nextIndex = (viewControllerIndex + 1) % pages.count`

DanilaG commented on Tracker/SplashViewController/OnboardingViewController.swift:
```diff
@@ -0,0 +1,120 @@
+        guard let viewControllerIndex = pages.firstIndex(of: viewController) else {
+            return nil
+        }
+        let previousIndex = viewControllerIndex - 1
+        
+        guard previousIndex >= 0 else {
+            return pages.last
+        }
```
Можно воспользоваться модульной арифметикой: `let nextIndex = (viewControllerIndex + pages.count - 1) % pages.count`

DanilaG commented on Tracker/CoreData/CoreDataStores/TrackerCategoryStore.swift:
Сейчас файл полностью закомментирован, давай уберём его из проекта)

DanilaG commented on Tracker/SplashViewController/TabBar/TrackersViewController.swift:
```diff
@@ -207,9 +212,42 @@ extension TrackersViewController {
+        collection.reloadData()
+    }
+    
+//    func updateTracker(with trackers: [Tracker], and category: String, _ id: UUID) {
+//        
+//        let categoryCoreDataID = categoryStore.fetchCategoryID(with: id)
+//        let categoryCoreData = categoryStore.fetchCategory(with: categoryCoreDataID)
+//        let tracker
+//        trackerStore.updateTracker(
+//            with: trackerStore.fetchTrackerID(with: trac),
+//            categoryCoreData: categoryCoreData,
+//            and: trackers.last!
+//        )
+//        categories.append(TrackerCategory(id: id,
+//                                          title: category,
+//                                          trackers: trackers))
+//    }
```
Давай уберём закоментированную  функцию, если она не нужна)

Peppa-Akk committed:
Sprint 16:
- Fix button (now it's not at the OnboardingVIewController);
- FIx selection of category.
